test_responses.py

```diff
@@ -390,3 +396,56 @@ def run():
 
     run()
     assert_reset()
+
+
+def test_response_timeout():
+    body = b'test'
+    status = 200
+    url = 'http://example.com/'
+    timeout = 0.1  # timeout in seconds
+
+    @responses.activate
+    def run():
+        responses.add(responses.GET, url, body=body, timeout=timeout)
+        begin = time.time()
+        resp = requests.get(url)
+        end = time.time()
+        assert end - begin >= timeout  # some additional delay may took place
+        assert resp.text == 'test'
+        assert resp.status_code == status
+
+    run()
+    assert_reset()
+
+
+def test_request_timeout():
+    status = 200
+    url = 'http://example.com/'
+    timeout = 0.1  # timeout in seconds
+
+    @responses.activate
+    def run():
+        responses.add(responses.GET, url, timeout=timeout)
+        begin = time.time()
+        resp = requests.get(url, timeout=timeout)
+        end = time.time()
+        assert end - begin >= timeout
+        assert resp.status_code == status
+
+    run()
+    assert_reset()
+
+
+def test_request_timeout_expired():
+    url = 'http://example.com/'
+    response_timeout = 0.2
+    request_timeout = 0.1
+
+    @responses.activate
+    def run():
+        responses.add(responses.GET, url, timeout=response_timeout)
```

Comment:
Actually, if you do that, then within the function you linked 
```
    def _on_request(self, adapter, request, **kwargs):
        match = self._find_match(request)
        # TODO(dcramer): find the correct class for this
        if match is None:
            error_msg = 'Connection refused: {0} {1}'.format(request.method,
                                                             request.url)
            response = ConnectionError(error_msg)
            response.request = request

            self._calls.add(request, response)
            raise response

        if 'body' in match and isinstance(match['body'], Exception):
            self._calls.add(request, match['body'])
            raise match['body']
```
`match['body']` will equal `ConnectTimeout` and therefore `isinstance(match['body'], Exception)` will be False.

You'll want to use `body=ConnectTimeout()` so that you actually pass an instance.